### PR TITLE
Specialize batched column-wise hashing for RunLengthEncodedBlock

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Adds special case handling of `RunLengthEncodedBlock` in generated `FlatHashStrategy` implementations that hoists the hash calculation outside of the loop. This has the added benefit of making the normal compute loop at most bi-morphic between `DictionaryBlock` and the type-corresponding `ValueBock` subclass.

`FlatGroupByHash` already specializes hash calculations for `RunLengthEncodedBlock` at a higher level when _all_ group by input blocks are run-length encoded, so this change will only affect performance for mutli-column grouping operations with a mix of run length and dictionary/plain block inputs.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java [hunk 1]
```diff
@@ -393,9 +396,13 @@ private static void generateHashBlocksBatched(ClassDefinition definition, List<K
...
             }
-            body.append(invokeStatic(method, blocks.getElement(keyField.index()), hashes, offset, length));
+            computeHashes.append(invokeStatic(method, blocks.getElement(keyField.index()), hashes, offset, length));
         }
-        body.ret();
+
+        body.append(new IfStatement("if (length != 0)")
+                .condition(equal(length, constantInt(0)))
+                .ifFalse(computeHashes))
+                .ret();
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java:408]: I think you can move this to be right after the checkFromIndexSize with an early return like this:
```
        body.append(new IfStatement("if (length == 0) return")
                        .condition(equal(length, constantInt(0)))
                        .ifTrue(RET));
```

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java [hunk 2]
```diff
@@ -429,35 +432,62 @@ private static MethodDefinition generateHashBlockVectorized(ClassDefinition defi
...
+                .append(mayHaveNull.set(block.invoke("mayHaveNull", boolean.class)))
+                .append(new ForLoop("for (int index = 0; index < length; index++)")
+                        .initialize(index.set(constantInt(0)))
+                        .condition(lessThan(index, length))
+                        .update(index.increment())
+                        .body(new BytecodeBlock()
+                                .append(new IfStatement("if (mayHaveNull && block.isNull(position))")
+                                        .condition(and(mayHaveNull, block.invoke("isNull", boolean.class, position)))
+                                        .ifTrue(hash.set(constantLong(NULL_HASH_CODE)))
+                                        .ifFalse(hash.set(computeHashNonNull)))
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: The change for `mayHaveNull` seems unrelated.  Also, you generally want to hoist that out of the loop and have two loops.

File: core/trino-main/src/test/java/io/trino/operator/TestFlatHashStrategy.java [hunk 3]
```diff
@@ -51,14 +51,12 @@
...
 {
-    private static final TypeOperators TYPE_OPERATORS = new TypeOperators();
-    private static final JoinCompiler JOIN_COMPILER = new JoinCompiler(TYPE_OPERATORS);
-
     @Test
     public void testBatchedRawHashesZeroLength()
     {
-        List<Type> types = createTestingTypes();
-        FlatHashStrategy flatHashStrategy = JOIN_COMPILER.getFlatHashStrategy(types);
+        TypeOperators typeOperators = new TypeOperators();
```
Comment (dain) [inline core/trino-main/src/test/java/io/trino/operator/TestFlatHashStrategy.java:57]: Is this point with moving these, to reduce testing memory or is there another goal.  If it just testing memory, you can make the fields instant fields, as Junit5 dereferences test classes properly (unlike TestNG)

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java [hunk 4]
```diff
@@ -466,7 +466,7 @@ private static MethodDefinition generateHashBlockVectorized(ClassDefinition defi
             rleHandling.append(new ForLoop("for (int index = 0; index < length; index++) { hashes[index] = CombineHashFunction.getHash(hashes[index], hash); }")
                     .initialize(index.set(constantInt(0)))
                     .condition(lessThan(index, length))
-                    .update(index.increment())
+                    .update(new BytecodeBlock().incrementVariable(index, (byte) 1))
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java:469]: I assume this means using IINC, actually matters (which seems strange given how HotSpot works), so can you update Variable and ByteCodeExpression instead, so all code gets this benefit?

**Review by dain — COMMENTED** [review]
Nice!

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java (hunk 2 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: The change is somewhat related. This is logically the same pattern that was present before special casing `RunLengthEncodedBlock`, we just assigned the value at the beginning of the method but were still using `mayHaveNull && block.isNull(position)` inside of the loop. So effectively, we've moved the assignment of `mayHaveNull` into the else-branch so that we don't need to check that for `RunLengthEncodedBlock`.

In either case, I chose to avoid splitting into two loops here to keep the bytecode size of the method small. Before C2 kicks in, we still benefit from branch prediction guessing the control flow reliably since the value of `mayHaveNull` does not change during the loop execution. After C2 kicks in, we're relying on the JIT to perform "loop unswitching" which effectively splits the single loop into two loops after removing the loop-invariant `mayHaveNull` check. You could do this explicitly in code of course, but it makes the code generator more complicated and the bytecode size significantly larger which affects early inlining behaviors (which are driven by bytecode size and not native code size).

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java:408]: Sure, done.

File: core/trino-main/src/test/java/io/trino/operator/TestFlatHashStrategy.java (hunk 3 above)
Comment (pettyjamesm) [reply core/trino-main/src/test/java/io/trino/operator/TestFlatHashStrategy.java:57]: Yep, will make these instance fields.

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java (hunk 4 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java:469]: It doesn't _really_ matter, but it is smaller in terms of bytecode size- `IINC` only works for `byte / short / int` but not `long` whereas the sequence of `variable = variable + 1` works regardless of size (but involves more bytecodes). I actually did this because I noticed that this was how you implemented the loop increments in `AccumulatorCompiler` which I assumed was on purpose on your behalf. I'll drop it from this PR and open the `IINC` logic against airlift.

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: airlift/bytecode PR: https://github.com/airlift/bytecode/pull/17

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: Turns out that sequence causes issues with the code generator because: `"JSR/RET are not supported with computeFrames option"`, so I've reverted back to the previous code shape instead.

Comment (dain) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: Oh well.  Something is broken with the library

**Review by dain — APPROVED** [review]
